### PR TITLE
Add a sensu_check resource, closes #10

### LINKED ISSUE #10: Sensu 2.0 check LWRP
Create a Sensu 2.0 check LWRP to manage Sensu checks via the Backend API.

Please see https://github.com/sensu/sensu-go-chef/issues/9 for implementation details.

### PR DESCRIPTION
**Pull request by webframp**
This will generate the object files necessary for a single check definition and
executes `sensuctl create -f` with the generated file. Includes `:create` and `:delete` actions. 

Includes unit/integration tests.

### CODE REVIEW

**Comment by webframp** [comment]
This is deliberately simple to start. 

Some ideas for future improvements:
- validate `cron` property syntax, ideally reusing what can be from builtin cron [resource](https://github.com/chef/chef/blob/master/lib/chef/resource/cron.rb) and [provider](https://github.com/chef/chef/blob/master/lib/chef/provider/cron.rb)
- check if `command` property exists
- validate if org/environment properties exist (sensuctl shell out)

File: resources/check.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+resource_name :sensu_check
+
+property :config_home, String, default: '/etc/sensu'
+property :command, String, default: '/bin/true', required: true
+property :cron, String, default: '* * * * *'
+property :subscriptions, Array, default: ['default']
+property :publish, [true, false], default: true
+property :handlers, Array, default: ['default']
```
Comment (majormoses) [inline resources/check.rb]: is there a default handler in sensu2? If not I don think we should default to anything.

File: test/cookbooks/sensu_test/recipes/default.rb [hunk 2]
```diff
@@ -5,3 +5,9 @@
 sensu_ctl 'default' do
   action [:install, :configure]
 end
+
+sensu_check 'cron' do
```
Comment (majormoses) [inline test/cookbooks/sensu_test/recipes/default.rb]: we should not assume cron is default and should specify it here.

File: test/integration/default/default_spec.rb [hunk 3]
```diff
@@ -41,3 +41,9 @@
   its('stdout') { should match /Username/ }
   its('exit_status') { should eq 0 }
 end
+
+describe command('sensuctl check list') do
```
Comment (majormoses) [inline test/integration/default/default_spec.rb]: we should use the as json matcher as this does not really give us the guarantee that the `dad_jokes` handler was added to `cron` and not some other check.

**Comment by majormoses** [comment]
IMO in order to close #10 we need to support every attribute the sensu check spec supports. I think for now we can hit the most important ones for now and add an `additional` that allows passing an arbitrary hash of key value pairs so that you can add non spec specific config such as a `runbook` link.

**Comment by webframp** [comment]
I added the rest of the documented attributes from: https://docs.sensu.io/sensu-core/2.0/reference/checks/#check-attributes but let me know if you see anything missing

File: resources/check.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+resource_name :sensu_check
+
+property :config_home, String, default: '/etc/sensu'
+property :command, String, default: '/bin/true', required: true
+property :cron, String, default: '* * * * *'
+property :subscriptions, Array, default: ['default']
```
[the remark this answers is not in the document]
Comment (webframp) [reply resources/check.rb]: works for me, I'll adjust

File: resources/check.rb (hunk 1 above)
Comment (webframp) [reply resources/check.rb]: sounds fine

File: resources/check.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+resource_name :sensu_check
+
+property :config_home, String, default: '/etc/sensu'
+property :command, String, default: '/bin/true', required: true
+property :cron, String, default: '* * * * *'
```
[the remark this answers is not in the document]
Comment (webframp) [reply resources/check.rb]: fine with me

File: libraries/helpers.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      spec = {}
+      spec['name'] = new_resource.name
+      spec['command'] = new_resource.command
+      spec['cron'] = new_resource.cron
+      spec['subscriptions'] = new_resource.subscriptions
+      spec['publish'] = new_resource.publish
+      spec['handlers'] = new_resource.handlers
+      spec['organization'] = new_resource.organization
+      spec['environment'] = new_resource.environment
+
```
[the remark this answers is not in the document]
Comment (webframp) [reply libraries/helpers.rb]: It's been added. Thanks

File: test/integration/default/default_spec.rb (hunk 3 above)
Comment (webframp) [reply test/integration/default/default_spec.rb]: Fixed

File: test/cookbooks/sensu_test/recipes/default.rb (hunk 2 above)
Comment (webframp) [reply test/cookbooks/sensu_test/recipes/default.rb]: added

File: resources/check.rb (hunk 1 above)
Comment (webframp) [reply resources/check.rb]: removed default

File: resources/check.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (webframp) [reply resources/check.rb]: removed default

File: resources/check.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (webframp) [reply resources/check.rb]: removed default value

**Comment by webframp** [comment]
Updated based on feedback. I agree about the tests being too generous, they're quite simple now and not comprehensive enough for what I'd like to see eventually

**Comment by webframp** [comment]
I'm not sure about something as generic as `additional` at this time. I think with sensu2 that's going to be handled by `extended_attributes` which can store arbitrary json. I'll go ahead and add extended_attributes and it can be used this way if someone wants.

**Comment by majormoses** [comment]
Makes sense I was just calling it `additional` because that is what the sensu 1.x cookbook called it. That being said I think I view them as potentially 2 different properties. Rationale:
- when a new check attribute is added if there is not some generic hash you can add keys to then it would require someone to come to the cookbook to add it and wait for review + release before they could leverage it. By having a generic hash they can immediately leverage it and then get it to become a first class property in the cookbook.
- extended attributes seems more like the generic add arbitrary metadata about the instance but I could be wrong as there is no real description yet in the docs.

**Comment by webframp** [comment]
I updated the properties that are tested and verified. Focused solely on inspec test at this point, chefspec testing will be mildly useful but I feel like it can be enhanced later.

I don't mind adding `additional` as a property if that's a blocker right now. I'm also a little unclear about it's relationship to `extended_attributes` and would probably need @portertech  to weigh in on it to be sure.

**Comment by majormoses** [comment]
I don't consider adding `additional` or whatever we call it to be a blocker just something we should be thinking about and put in shortly. Let's hold off on that until we understand the relationship of `extended_attributes` as I don't think it serves a different use case but I'd like to confirm.

**Comment by majormoses** [comment]
I have opened #16 to address the questions around `extended_attributes` and how we want to allow people to add keys before they exist in the cookbook.

**Pull request merged**